Adding SourceFluxStatistics

MelReyCG opened the pull request:
The goal of this PR is to add a `SourceFluxStatistics` component which aggregate the producted mass & rate on a `SourceFlux` set.
- Log Level 1 outputs the sum of all `SourceFluxStatistics` produced rate & mass,
- Log Level 2 details values for each SourceFluxStatistics,
- Log Level 3 details values for each region.

Here is the typical log output for a multi-phase simulation, with a log level of 2 :
```
Rank 0: SourceFluxStatistics timeStepFluxStats (of sourceFlux, in all_regions): Producing on 1 elements
Rank 0: SourceFluxStatistics timeStepFluxStats (of sourceFlux, in all_regions): Produced mass = [-280.5, 0] kg
Rank 0: SourceFluxStatistics timeStepFluxStats (of sourceFlux, in all_regions): Production rate = [-0.561, 0] kg/s
Rank 0: SourceFluxStatistics timeStepFluxStats (of sinkFlux, in all_regions): Producing on 1 elements
Rank 0: SourceFluxStatistics timeStepFluxStats (of sinkFlux, in all_regions): Produced mass = [0, 31] kg
Rank 0: SourceFluxStatistics timeStepFluxStats (of sinkFlux, in all_regions): Production rate = [0, 0.062] kg/s
Rank 0: SourceFluxStatistics timeStepFluxStats (of flux_set, in all_regions): Producing on 2 elements
Rank 0: SourceFluxStatistics timeStepFluxStats (of flux_set, in all_regions): Produced mass = [-280.5, 31] kg
Rank 0: SourceFluxStatistics timeStepFluxStats (of flux_set, in all_regions): Production rate = [-0.561, 0.062] kg/s
```

Note: If the `PeriodicEvent` that execute the `SourceFluxStatistics` has a frequency such as multiple time-step are encountered, the statistics are aggregated (masses -> sum, rates -> mean).


I've also added a test which checks if the statistics output of the `SourceFluxStatistics` and `SinglePhaseStatistics::RegionStatistics::totalMass` are consistent with the provided production rates table.
The test checks if the statistics are right over the whole simulation, but also for each timesteps. It will crash if a timestep does not run. Thus, it tests if the `PeriodicEvent` functions as expected (is it the only test to do so ?).

MelReyCG committed:
added region mass stats

MelReyCG committed:
added region temperature stats

MelReyCG committed:
Logging debug info - to be removed laster

MelReyCG committed:
better debug output

MelReyCG committed:
annotations

MelReyCG committed:
cleaning

MelReyCG committed:
Output flux stats in log

MelReyCG committed:
completed previous commit

MelReyCG committed:
Add the logLevel viewkey & description

MelReyCG committed:
Add a way to expend the description of a wrapper

MelReyCG committed:
Intermediate progress to properly store the SourceFlux stats

MelReyCG committed:
adding the SourceFluxStatistics Task which will allow to aggregate statistics of SourceFluxBoundaryCondition (WIP)

MelReyCG committed:
simplify description to avoid regenerating a lot of .rst (to be added in a more specific PR later)

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into feature/rey/sourceflux-stats

MelReyCG committed:
missed a few details

MelReyCG committed:
forgotten generated files...

MelReyCG committed:
SourceFlluxStatistics now compiles

MelReyCG committed:
generated docs

MelReyCG committed:
SourceFluxStatistics now runs:
- finalize architecture
- add a way to accumulate statistics over multiple timesteps
- test case runs correctly
- added a way to know the region of a subregion
- cleanup
- completed some wrappers descriptions

MelReyCG committed:
few edits

MelReyCG committed:
uncrustify

MelReyCG committed:
architecture adaptations to expose values to unit tests

MelReyCG committed:
added singlephase test

MelReyCG committed:
removed debug infos

MelReyCG committed:
unit test: code layout, removed debug info

MelReyCG committed:
unit test: code layout, removed debug info, xmlInput improvements

MelReyCG committed:
Merge branch 'feature/rey/sourceflux-stats' of https://github.com/GEOS-DEV/GEOS into feature/rey/sourceflux-stats

MelReyCG committed:
Added multiphase xml input

MelReyCG committed:
few comments

MelReyCG committed:
code style

MelReyCG committed:
unit test: Added checks on every timesteps stats

MelReyCG committed:
unit test: separated the input parameters and expected value set

MelReyCG committed:
unit test: cleaning comments

MelReyCG committed:
unit test: compositional multiphase sim now runs

MelReyCG committed:
allowed multiple component stats for SourceFluxStatistics

MelReyCG committed:
changed singlephase parameters so that they are not multiple among themselves

MelReyCG committed:
forgot to replace some hard coded values

MelReyCG committed:
Whole multiphase sim stat check

MelReyCG committed:
crash investigating...

MelReyCG committed:
fixed crashed :)

MelReyCG committed:
Revert unwanted pushed files
This reverts commit e2ac4465356fa6c2d8f4d640a17fed02202e3c4c.

MelReyCG committed:
Finished the multiphase test

MelReyCG committed:
Unexpected result debugging
-> bug was due to not taking into account a timestep could "restart" as it can be cut when reaching maxNewtonIter.

MelReyCG committed:
forgotten decls

MelReyCG committed:
code cleaning

MelReyCG committed:
fixing gpu captures + adding support for ReactiveCompositionalMultiphaseOBL

MelReyCG committed:
fixed sum operations (=> reduceSum)

MelReyCG committed:
uncrustify

MelReyCG committed:
test pvt tables optimisation (13.4s -> 3.0s)

MelReyCG committed:
exposed SinglePhaseStatistics::RegionStatistics

MelReyCG committed:
compared mass difference from SinglePhaseStatistics and SourceFluxStatistics

MelReyCG committed:
Merge branch 'develop' into feature/rey/sourceflux-stats

MelReyCG committed:
fixed typo

MelReyCG committed:
Merge branch 'feature/rey/sourceflux-stats' of https://github.com/GEOS-DEV/GEOS into feature/rey/sourceflux-stats

paveltomin commented:
please double check that logic is consistent with useMass flag (default is 0)

paveltomin commented:
> please double check that logic is consistent with useMass flag (default is 0)

and time step chop

MelReyCG committed:
Added guard to test sub-timesteps

MelReyCG commented:
@paveltomin

Timestep cuts were tested by the `checkMultiPhaseFluxStatistics` test (this simulation has 3 sub-timesteps). I added a guard (`testInputs.requireSubTimeStep`) so it will keep being tested with timesteps.

About `useMass`, the `checkMultiPhaseFluxStatistics` simulation does not converge anymore when I set it to 1. I am probably missing how to manage this component. Are the `SourceFlux` units in mole by default ? I'll investigate tomorrow.

Side question : do you think I should add CSV export for this component ?

paveltomin commented:
> About `useMass`, the `checkMultiPhaseFluxStatistics` simulation does not converge anymore when I set it to 1. I am probably missing how to manage this component. Are the `SourceFlux` units in mole by default ? I'll investigate tomorrow.

That's the tricky part, up to my understanding SourceFlux units depend on useMass flag. If useMass=0 then units are moles, if useMass=1 then units are mass.

> Side question : do you think I should add CSV export for this component ?

I found CSV output for flow and mechanics very convenient and use it all the time, also I think @jhuang2601 started using it. It would be nice to have it for SourceFlux too yes, but it can be done in a separate PR.

paveltomin commented:
> Timestep cuts were tested by the `checkMultiPhaseFluxStatistics` test (this simulation has 3 sub-timesteps). I added a guard (`testInputs.requireSubTimeStep`) so it will keep being tested with timesteps.

You test one new code with another new code :)

Your current implementation is tricky and only valid for constant flux value because if I understood correctly the statistics is being accumulated over the newton iterations while the real value that matters is the final one, when the iterations converged.
I don't see any zeroing-out of accumulated quantities when time step cut happens but I am probably blind. My suspicion for time step cut is similar - only works because the flux value is constant, but fundamental logic behind is slippery.

MelReyCG committed:
Added CSV export & more docs

MelReyCG commented:
@paveltomin 

I've added the CSV export and try to make the code of `SourceFluxStatsAggregator::WrappedStats::gatherTimeStepStats()` clearer. Do you think this method is missing something about timestep cuts? Or do you think I should go for another way to keep track of data accumulation.

> You test one new code with another new code :)

I did not understand if you are thinking that is a good thing.
I made a typo it the comment you quoted -> What I did was to add a check in my unit test to ensure that sub-timesteps are being tested.

paveltomin commented:
> I've added the CSV export and try to make the code of `SourceFluxStatsAggregator::WrappedStats::gatherTimeStepStats()` clearer. Do you think this method is missing something about timestep cuts? Or do you think I should go for another way to keep track of data accumulation.

Thanks for CSV and if you say timestep cut is ok, I trust that, thanks for checking and refining the code

> > You test one new code with another new code :)
> 
> I did not understand if you are thinking that is a good thing. I made a typo it the comment you quoted -> What I did was to add a check in my unit test to ensure that sub-timesteps are being tested.

I just mean that one could potentially do the same mistake in the main code and in the test code, so unit test may not help catching it

Would you mind adding that output into one of the xmls?

MelReyCG commented:
@paveltomin 

> I just mean that one could potentially do the same mistake in the main code and in the test code, so unit test may not help catching it

Oh I see!
To explain my test, I prepared some expected data and a xml input and check that the simulation outputs statistics that are consistent with the expected ones, so the test is not dependent of the internal code. I also chek if the count of timestep is as expected. And now I check if at least one timestep cut occurs, in order to see if everything goes right in that case too.

> Would you mind adding that output into one of the xmls?

Which XMLs output are you talking about ?

paveltomin commented:
> > Would you mind adding that output into one of the xmls?
> 
> Which XMLs output are you talking about ?

I mean add this new SourceFluxStatistics output to one of the existing XMLs

MelReyCG committed:
Removing the MultifluidBase::useMass wrapper as it is set by the solver anyway (by setMassFlag() ).

MelReyCG committed:
fixed catalog name of TimeStepChecker

MelReyCG commented:
Ok I will add one or more `SourceFluxStatistics` in the exemples.

`useMass` has indeed an influence on the `SourceFlux` units 👍 :
- I fixed the `SourceFluxStatistics` and added a (temporary) test to compare with mass and mole,
- I added descriptions of the wrapper about the fluxes units.
- I added a way of getting the *amount of fluid unit* on `FlowSolverBase`, and uniformized a bit with the other statistics classes.
- I verified the `SinglePhaseStatistics` units.

I don't know why, but the same test simulation in mass and mole are behaving differently (`MultiPhaseFluxStatisticsTestMol` has a timestep cut whereas `MultiPhaseFluxStatisticsTestMass` haven't any), here is the log :
[test2.txt](https://github.com/GEOS-DEV/GEOS/files/14361746/test2.txt)

paveltomin commented:
> I don't know why, but the same test simulation in mass and mole are behaving differently (`MultiPhaseFluxStatisticsTestMol` has a timestep cut whereas `MultiPhaseFluxStatisticsTestMass` haven't any), here is the log : [test2.txt](https://github.com/GEOS-DEV/GEOS/files/14361746/test2.txt)

this is normal, since the amount injected can be very different, solver can be more stressed

MelReyCG commented:
Ok I see. Thanks a lot for your inputs and review !

MelReyCG committed:
Fixed solver "amount of fluid" units (kg or mol), and added documentation to clarify its usage.

MelReyCG committed:
better timestep cut checking + refactored a bit

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into feature/rey/sourceflux-stats

MelReyCG committed:
generated docs

MelReyCG committed:
uncrustify

MelReyCG committed:
typo

MelReyCG committed:
reverting undesired paste

MelReyCG committed:
wrappers like this one is required to be able to "deliverClone()"

MelReyCG committed:
generated documentations

MelReyCG committed:
docs

MelReyCG committed:
wrong doxygen keyword

MelReyCG committed:
doxygen test

MelReyCG committed:
test2

MelReyCG committed:
fixed doxygen

MelReyCG committed:
ajusting test value to facilitate convergence

MelReyCG committed:
code reordering + docs

MelReyCG committed:
debugging sub-sub-timestep behaviour

MelReyCG committed:
finally fixed numerical precision issue

MelReyCG committed:
removing double decl

MelReyCG committed:
removing unrelated code change

MelReyCG committed:
ajusting test value to facilitate convergence

MelReyCG committed:
test

MelReyCG committed:
allow non converged solution is tests as trilinos struggles with the testFluidStatistics

paveltomin commented on src/coreComponents/unitTests/fluidFlowTests/testFluidStatistics.cpp:
```diff
@@ -0,0 +1,1199 @@
+                                 newtonMaxIter="8"
+                                 allowNonConverged="1" />
+      <LinearSolverParameters solverType="gmres"
+                              preconditionerType="amg"
```
should be mgr i guess

MelReyCG committed:
Solver setting test

MelReyCG committed:
Solver setting test 2 (Mass test is passed with trilinos)

MelReyCG committed:
Solver setting test

MelReyCG committed:
reverted to previous settings

MelReyCG committed:
forgotten ILUK attribute

MelReyCG committed:
trying to test sub-timesteps only with hypre

MelReyCG committed:
trilinos & hyper work in stastifying time

MelReyCG committed:
removed unused headers

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into feature/rey/sourceflux-stats

MelReyCG committed:
solver precision

MelReyCG committed:
avoiding exact 0 / 1 values in test

paveltomin commented on src/coreComponents/unitTests/fluidFlowTests/testFluidStatistics.cpp:
```diff
@@ -0,0 +1,1197 @@
+/*
```
could you please rename it to test**Flow**Statistics.cpp?

MelReyCG replied:
It is indeed a better name

paveltomin commented on src/coreComponents/unitTests/fluidFlowTests/testFluidStatistics.cpp:
```diff
@@ -0,0 +1,1197 @@
+
+      <NonlinearSolverParameters newtonTol="1.0e-6"
+                                 newtonMaxIter="40"
+                                 allowNonConverged="1" />
```
remove

MelReyCG committed:
removed parameter as they are to defualt values

MelReyCG committed:
renaming testFluidStatistics to testFlowStatistics

dkachuma commented on src/coreComponents/fieldSpecification/SourceFluxStatistics.hpp:
This looks like a task that should belong to `physicsSolvers` and not quite a `fieldSpecification`.

MelReyCG replied:
So the solvers would expose the producted rates & masses for each `SourceFlux`, and then the `SourceFluxStatistics` would access to it?

dkachuma commented on src/coreComponents/fieldSpecification/SourceFluxStatistics.cpp:
```diff
@@ -0,0 +1,364 @@
+    appendDescription( GEOS_FMT( "\n- Log Level 1 outputs the sum of all {0}(s) produced rate & mass,\n"
+                                 "- Log Level 2 details values for each {0},\n"
+                                 "- Log Level 3 details values for each region,\n"
+                                 "- Log Level 4 details values for each sub-region.",
```
I wouldn't go to the sub-region level. I would stop at the region level.

dkachuma commented on src/coreComponents/fieldSpecification/SourceFluxStatistics.cpp:
```diff
@@ -0,0 +1,364 @@
+    {
+      registerWrappedStats( mesh, fluxName, viewKeyStruct::allRegionWrapperString() );
+
+      mesh.getElemManager().forElementRegions( [&]( ElementRegionBase & region )
```
If the `logLevel` is low enough, then we don't need to create these wrappers.

MelReyCG replied:
I use them to collect stats from the `SinglePhaseBase::applySourceFluxBC()` methods that have no knowledge of the regions. Can I keep them to collect the data or should I proceed differently?

TotoGaz commented on src/coreComponents/fieldSpecification/SourceFluxStatistics.cpp:
```diff
@@ -0,0 +1,364 @@
+  }
+}
+
+bool SourceFluxStatsAggregator::execute( real64 const GEOS_UNUSED_PARAM( time_n ),
+                                         real64 const GEOS_UNUSED_PARAM( dt ),
+                                         integer const GEOS_UNUSED_PARAM( cycleNumber ),
+                                         integer const GEOS_UNUSED_PARAM( eventCounter ),
+                                         real64 const GEOS_UNUSED_PARAM( eventProgress ),
+                                         DomainPartition & domain )
```
Hi, I'd be interested in decoupling the gathering/aggregation of the data from its output.
Here we impose `log` and `csv`, for example. There's no reason not to use the `CSV` for other types of data, and to impose `CSV` for the source flux data.

@MelReyCG , I had a quick chat with @arng40 about that.

TotoGaz commented on src/coreComponents/fieldSpecification/SourceFluxStatistics.cpp:
```diff
@@ -0,0 +1,364 @@
+  }
+}
+
+bool SourceFluxStatsAggregator::execute( real64 const GEOS_UNUSED_PARAM( time_n ),
```
Hi, I'd be interested in decoupling the gathering/aggregation of the data from its output.
Here we impose `log` and `csv`, for example. There's no reason not to use the `CSV` for other types of data, and to impose `CSV` for the source flux data.

@MelReyCG , I had a quick chat with @arng40 about that.

MelReyCG replied:
Just to notify everyone about the conclusion of our chat on this subject:
- There are currently a lot of these log / CSV output in the statistics classes,
- Statistics output must be unified, and not strictly bound to CSV,
- This output refactoring is not the subject of this PR, but **is a task that I have got in my roadmap**.
- An `Outputs` new node could be created to output those stats, maybe something like:
```XML
<Outputs>
  <StatsOutput
    name="mySourceFluxesOutput"
    source="mySourceFluxesStats"
    outputFile="sourceFluxStats.csv" /> <!-- could be json or so -->
</Outputs>
```

TotoGaz reviewed (commented):
That looks like a nice feature.
A question about the display though.

MelReyCG committed:
removing the 4th log level

MelReyCG committed:
missing inlines

MelReyCG committed:
xsd / rst changes

MelReyCG committed:
moving the TimeStepChecker in its own file (& created a testingUtilities folder)

MelReyCG committed:
Merge remote-tracking branch 'origin/develop' into feature/rey/sourceflux-stats

MelReyCG committed:
moving some definition to cpp

MelReyCG committed:
uncrustify :s

CusiniM commented on src/coreComponents/common/Units.hpp:
```diff
@@ -122,6 +138,14 @@ constexpr inline std::string_view getSymbol( Unit unit )
     case TemperatureInC:  return "C";
     case Distance:        return "m";
     case Time:            return "s";
+    case Viscosity:       return "Pa*s";
+    case Enthalpy:        return "J/kg";
+    case Density:         return "kg/m3";
+    case Solubility:      return "g/L";
+    case Mass:            return "kg";
+    case Mole:            return "mol";
+    case MassRate:        return "kg/s";
+    case MoleRate:        return "mol/s";
```
Units are not really strongly enforced in GEOS. One could use any set of units as long as they are consistent. For example a user can decide to specify pressures in MPa an adjust all other properties. I would say that we either properly introduce a unit parser or these units can easily be false.

MelReyCG replied:
About this header, it is in S.I. by default (following the GEOS convention). Some units can be in other systems as you can see for the temperatures (°C rather than °K), and if someone want to add more, I think it should come with conversion functions.
I strongly advise to add comments on all non-S.I. variables / method, it would be less error-prone and would help when implementing a parsing unit system or when adding new outputs.

About adding unit parsing system, it is on my roadmap. Feel free to add suggestions on [my EPIC](https://github.com/GEOS-DEV/GEOS/issues/2704).

TotoGaz commented on src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseStatistics.cpp:
```diff
@@ -305,7 +304,7 @@ void CompositionalMultiphaseStatistics::computeRegionStatistics( real64 const ti
                                         subRegionComponentMass.toView() );
 
-    ElementRegionBase & region = elemManager.getRegion( subRegion.getParent().getParent().getName() );
+    ElementRegionBase & region = elemManager.getRegion( ElementRegionBase::getParentRegion( subRegion ).getName() );
```
That's better!

TotoGaz commented on src/coreComponents/unitTests/testingUtilities/TestingTasks.hpp:
```diff
@@ -0,0 +1,80 @@
+ * As this Group is designed for developpers and not user oriented, it should not appear in the documentation nor in the xsd.
+ * Question: could this task be used in the integratedTests ?
+ */
+class TimeStepChecker : public TaskBase
```
@CusiniM What do you think of this. I believe that's interesting.

TotoGaz commented on src/coreComponents/physicsSolvers/SolverBase.hpp:
```diff
@@ -734,6 +734,7 @@ class SolverBase : public ExecutableGroup
   virtual bool registerCallback( void * func, const std::type_info & funcType ) final override;
 
   SolverStatistics & getSolverStatistics() { return m_solverStatistics; }
+  SolverStatistics const & getSolverStatistics() const { return m_solverStatistics; }
```
Who's using the non-const version?
If you can, can you remove it or use a setter instead?

MelReyCG replied:
It is used at `src/coreComponents/physicsSolvers/multiphysics/CoupledSolver.hpp:431, 491, 506`, I think the non-const version is useful, should I remove the const one?

TotoGaz commented on src/coreComponents/fieldSpecification/SourceFluxStatistics.cpp:
```diff
@@ -0,0 +1,361 @@
+  Base::postProcessInput();
+
+  FieldSpecificationManager & fsManager = FieldSpecificationManager::getInstance();
+  if( m_fluxNames.size() == 1 && m_fluxNames[0] == "all" )
```
```suggestion
  if( m_fluxNames.size() == 1 && m_fluxNames[0] == "*" )
```
Even if we're not doing the `fnmatch` in this PR, would using `*` make sense?

MelReyCG replied:
I agree, these `"all"` proposals came to my mind as it is possible to use that as `setNames` in field specifications.

TotoGaz commented on src/coreComponents/fieldSpecification/SourceFluxStatistics.cpp:
```diff
@@ -0,0 +1,361 @@
+        regionStats.stats() = StatData();
+
+        forAllSubRegionStatsWrappers( region, regionStats.getFluxName(),
+                                      [&] ( ElementSubRegionBase &, WrappedStats & subRegionStats )
```
4 nested lambda functions, I think that you tie the record 🤣 😉

TotoGaz commented on src/coreComponents/fieldSpecification/SourceFluxStatistics.cpp:
```diff
@@ -0,0 +1,361 @@
+  }
+
+  // combine period results from all MPI ranks
+  m_stats.mpiReduce();
```
You need to be sure that this is called by all the ranks. Is it the case?

MelReyCG replied:
Indeed, I tested that all ranks send & combine how much the `SourceFlux` produced / injected (possibly 0).

TotoGaz commented on src/coreComponents/fieldSpecification/SourceFluxStatistics.hpp:
```diff
@@ -0,0 +1,420 @@
+                              arrayView1d< real64 const > const & producedMass,
+                              integer elementCount );
+
+    /**
+     * @brief Finalize the period statistics of each timestep gathering and render data over all mpi ranks.
+     * The result can be read by the data() assessor.
+     */
```
Be a little more explicit about the `MPI` call that needs to be done by everyone. Maybe add a `@note` ?

TotoGaz commented on src/coreComponents/fieldSpecification/SourceFluxStatistics.hpp:
```diff
@@ -0,0 +1,420 @@
+    constexpr inline static string_view fluxSetWrapperString() { return "flux_set"; }
+  };
+
+protected:
+
+  /**
+   * @copydoc Group::postProcessInput()
+   */
+  void postProcessInput() override;
```
Can you move this `protected` function between the `public` and the `private`?
+ Are you sure that you need it to be `protected`?

MelReyCG replied:
I set the class as `final` and removed the protected qualificator.

TotoGaz commented on src/coreComponents/fieldSpecification/SourceFluxStatistics.hpp:
```diff
@@ -0,0 +1,420 @@
+#include "physicsSolvers/FieldStatisticsBase.hpp"
+#include "physicsSolvers/fluidFlow/FlowSolverBase.hpp"
+
+class SourceFluxBoundaryCondition;
```
Do you need that?

MelReyCG replied:
Oops, I don't need this forward declaration anymore.

MelReyCG committed:
few adjustments

MelReyCG committed:
Merge commit '79856a0e653a508b4fe4dcd8c582fcdc2838b756' into feature/rey/sourceflux-stats

MelReyCG committed:
added an header due to incomplete DomainPartition type

MelReyCG committed:
by default, SourceFluxStat targets al SourceFlux + adapted test

MelReyCG committed:
generated files

MelReyCG committed:
Added star character * in group name reference regex.

paveltomin committed:
Merge branch 'develop' into feature/rey/sourceflux-stats

MelReyCG commented:
I am currently working on the parallel implementation, so please wait before merging this PR.

paveltomin commented:
> I am currently working on the parallel implementation, so please wait before merging this PR.

I moved it into "Approved" to avoid confusion

MelReyCG commented:
@paveltomin Sorry, I posted on the wrong PR, I was talking about the transmissibility one, this one works fine!

paveltomin committed:
Merge branch 'develop' into feature/rey/sourceflux-stats

paveltomin committed:
Merge branch 'develop' into feature/rey/sourceflux-stats
